pkg/sbom/spdx/marshal.go

```diff
@@ -389,11 +400,96 @@ func (m *Marshaler) spdxAnnotations(c *core.Component, timeNow string) []spdx.An
 	return annotations
 }
 
-func (m *Marshaler) spdxLicense(c *core.Component) string {
+func (m *Marshaler) spdxLicense(c *core.Component) (string, []*spdx.OtherLicense) {
+	// Only library components contain licenses
+	if c.Type != core.TypeLibrary {
+		return "", nil
+	}
 	if len(c.Licenses) == 0 {
-		return noAssertionField
+		return noAssertionField, nil
+	}
+	return m.normalizeLicenses(c.Licenses)
+}
+
+func (m *Marshaler) normalizeLicenses(licenses []string) (string, []*spdx.OtherLicense) {
+	var otherLicenses = make(map[string]*spdx.OtherLicense) // licenseID -> OtherLicense
+
+	// Save text licenses as OtherLicense
+	for i, license := range licenses {
+		if strings.HasPrefix(license, licensing.LicenseTextPrefix) {
+			otherLicense := m.newOtherLicense(strings.TrimPrefix(license, licensing.LicenseTextPrefix), true)
+			otherLicenses[otherLicense.LicenseIdentifier] = otherLicense
+			licenses[i] = otherLicense.LicenseIdentifier
+		}
+	}
+
+	license := strings.Join(lo.Map(licenses, func(license string, index int) string {
+		// e.g. GPL-3.0-with-autoconf-exception
+		license = strings.ReplaceAll(license, "-with-", " WITH ")
+		license = strings.ReplaceAll(license, "-WITH-", " WITH ")
+		return fmt.Sprintf("(%s)", license)
+	}), " AND ")
+
+	normalizedLicense, err := expression.Normalize(license, licensing.NormalizeLicense, expression.NormalizeForSPDX)
```

Comment:
I wanted to override the `expression.NormalizeForSPDX` function, but we use `With` as a delimiter to parse licenses:
https://github.com/aquasecurity/trivy/blob/6472e3c9da2a8e7ba41598a45c80df8f18e57d4c/pkg/licensing/expression/parser.go.y#L14
So there are problems with overwriting licenses with the wrong exception